### ISSUE TITLE
Dotted pattern should fail if argument is not inferrable

### ISSUE THREAD

**Issue by gallais** [agda/agda#3910]
In 2.5.3 the following module fails to typecheck, it is now accepted

```agda
open import Agda.Builtin.Nat

dottest : Nat → Nat
dottest ._ = {!!}

-- 2.5.3
-- Failed to infer the value of dotted pattern
-- when checking that the pattern _ has type Nat

-- 2.6.1
-- Accepted
```

**Comment by andreasabel** [comment]
I guess this feature got lost in @jespercockx big redesign of dot-pattern handling. I cannot judge if it is easy to bring back.
```agda
dottest : Nat → Nat
dottest .zero = {!!}
```
gives
```
zero != x of type Nat
when checking that the given dot pattern zero matches the inferred
value x
```
(a slightly questionable error message since the user did not write `x`, and it is not indicated that this is a system-generated identifier. When I try to please Agda's complaints in the literal sense,
```agda
dottest : Nat → Nat
dottest .x = {!!}
```
Agda is of course not happy either:
```
Not in scope: x
when scope checking x
```
)

Since Agda still seems to care about correctness of dot patterns, maybe it is sufficient to bring some distinction between `._` and `_` back?

**Comment by andreasabel** [comment]
It seems if a dot pattern has (by position) de Bruijn index `i` and is during unification solved by `v`, then `v` should not contain `i`.  More generally, de Bruijn indices corresponding to dot patterns should not occur in the dot pattern solutions, no?  Currently there is no such test, this is the reason for this issue.

A naive implementation of such a test however fails because Agda is careless when unifying two de Bruijn indices, which to instantiate by which.  Previously there was a heuristic [1] that would prefer the replacement of indices corresponding to the dot patterns rather than the indices corresponding to pattern variables, but this heuristic was retired.

Couldn't we setup unification such that it does not make eager decisions which index to assign, but rather keeps a union-find kind of structure to manage identification of indices?  At the very end of splitting, when we see the final picture, we could then make an informed decision.

[1] https://github.com/agda/agda/blob/6f25159f7e5e2a540ea6e1e02b9fff08b8c2576e/src/full/Agda/TypeChecking/Rules/LHS.hs#L94-L99

**Comment by jespercockx** [comment]
I'm away from my pc until Tuesday, but I just want to add that `x@.x` is currently a valid pattern.

**Comment by andreasabel** [comment]
Isn't this a bit odd?  Shouldn't `foo x@.x = rhs` (which is accepted) be sugar for `foo .x = let x = x in rhs`(which is not accepted)?

**Comment by jespercockx** [comment]
I agree that it's somewhat odd, but it's more of a stylistic choice than a foundational one. I'm not fundamentally opposed to changing how dot patterns work from how it is implemented currently, as long as there is a clear and simple semantics. Currently, the semantics of dot expressions is that:

1. They should be valid expressions in the context of the RHS of the clause (including hidden/implicit variables)
2. The pattern computed internally by case splitting should be an instance of the pattern written by the user (excluding placement of dot patterns and `@` patterns).

If you propose a different semantics that excludes the cases mentioned in this issue, we could adopt that one instead.

**Comment by jespercockx** [comment]
> a slightly questionable error message since the user did not write x, and it is not indicated that this is a system-generated identifier

This would be solved with #3215

**Comment by andreasabel** [comment]
The change would be that dot-patterns do not bind a variable, unlike variable patterns, underscores, and hidden (thus, omitted) patterns.  This would modify 1., which does not say explicitly now what the bound variables actually are, i.e., how this context is computed.

**Comment by jespercockx** [comment]
With the current infrastructure for LHS checking, all arguments are considered bound whether you want it or not. So if you want the behaviour you describe, we need to introduce a new kind of splitting operation that explicitly brings a variable out of scope (i.e. a weakening). Then, when Agda encounters a dot pattern it could use this new operation instead of considering splitting to be finished (as it does now).
